### PR TITLE
Improve error message to access balena settings

### PR DESCRIPTION
**Pull request by josecoelho**
Resolves: #1667  
Change-type: patch 
Depends-on:  https://github.com/balena-io-modules/balena-errors/pull/47 and https://github.com/balena-io-modules/balena-settings-storage/pull/29

- [x] Blocked until new version of balena-sdk is properly generated

### Windows Error message:
```
You don't have sufficient privileges to access balena settings.

Balena stores the session token in C:\Users\<user>\_balena.
This error usually indicates that the user doesn't have permissions over that directory,
which can happen if balena was executed as the root user.

Try resetting the ownership by opening a new Command Prompt as administrator and running: `takeown /f %USERPROFILE%\_balena /r`

Additional information may be available with the `--debug` flag.
For help, visit our support forums: https://forums.balena.io
For bug reports or feature requests, see: https://github.com/balena-io/balena-cli/issues/
```

### UNIX error message:
```
You don't have sufficient privileges to access balena settings.

Balena stores the session token in $HOME/.balena.
This error usually indicates that the user doesn't have permissions over that directory,
which can happen if balena was executed as the root user.

Try resetting the ownership by running: `sudo chown -R $(whoami) $HOME/.balena`

Additional information may be available with the `--debug` flag.
For help, visit our support forums: https://forums.balena.io
For bug reports or feature requests, see: https://github.com/balena-io/balena-cli/issues/
```

### CODE REVIEW

File: lib/errors.ts
Comment (josecoelho) [inline lib/errors.ts:125]: I'm not sure if we need to write tests for it.

Comment (pdcastro) [reply lib/errors.ts:125]: I think testing manually would be OK in this case. The CLI's `package.json` mentions both `balena-errors` and `balena-settings-storage`, so may need updating after balena-io-modules/balena-settings-storage/pull/29 is merged.

Comment (josecoelho) [reply lib/errors.ts:125]: I did another manual test after updating my branch to the latest master and I am getting 
`Login required: use the “balena login” command to log in.` instead of the custom error message. I'm not sure why.

I will investigate it further later.

Comment (josecoelho) [reply lib/errors.ts:125]: `balena-auth` still with the old version of `balena-settings-storage`. That's why.

Comment (pdcastro) [reply lib/errors.ts:125]: Check this PR's diff for the `npm-shrinkwrap.json` file: while the CLI is using `balena-settings-storage` v7, `balena-auth` still requires `balena-settings-storage` v6. Both versions v6 and v7 will be found in the CLI's `node_modules` folder, and `balena-auth` will use v6, not v7. Perhaps a pull request is also needed on the `balena-auth` repo, to update its `package.json`: https://github.com/balena-io-modules/balena-auth/blob/v4.0.2/package.json#L66

**Comment by josecoelho** [comment]
The build is expected to fail. It should be ok after all dependent PRs are merged.

File: package.json
Comment (pdcastro) [inline package.json:211]: This bump is welcome because we want and need to keep the CLI updated with changes to the SDK, but it also means that we will have to make some (manual\*) testing of the CLI with reference to the SDK's Changelog:
https://github.com/balena-io/balena-sdk/blob/master/CHANGELOG.md

\* Manual testing because the automated test coverage in the CLI is still limited.

Comment (josecoelho) [reply package.json:211]: Is there a list of manual tests that usually needs to happen? I could run it it that helps.
Or is it about going through the changes of `balena-sdk` and make sure all possible affected areas of the `cli` are working as expected?

Are you planning to add automated tests in relation to these dependencies? It's probably not related to this issue.

I can write an automated test for the current scenario, that would pave the way for other integration tests. This case is tricky, because of the root user situation, but we can still try to mock `fs` itself. Or we could try changing CI to not run as `root`. What do you reckon?

Comment (pdcastro) [reply package.json:211]: It is about going through the changes of `balena-sdk` and making sure all likely affected areas of the CLI are working as expected - sadly. 😞 

Automated tests for the current scenario are welcome, however I have a feeling (unconfirmed) that the current scenario is less likely to fail than other areas of the CLI that may be affected by the `balena-sdk` bump. Which would mean that the best use of time / effort, in relation to testing, would be to write test cases for CLI commands that use the SDK and don't currently have any tests at all. Of course that while this would be welcome, I would not expect you to do this as it is not related to issue #1667 which is the topic of this PR.

Another issue is that [versionbot failed to produce balena-sdk version 15.18.0](https://github.com/balena-io/balena-sdk/pull/1018#issuecomment-729236877). Until it gets sorted, this PR could be tested with the following line in `package.json`:
```
  "balena-sdk": "^15.18.0-josecoelho-update-balena-auth-36282cee7b3f394b690d0d430dc5705c075b67d1",
```

> This case is tricky, because of the root user situation, but we can still try to mock `fs` itself. Or we could try changing CI to not run as `root`. What do you reckon?

I think that mocking `fs` is a better option for now. Changing the CI to not run as `root` would be a large endeavour given our CI infrastructure (macOS and Windows workers run on physical hosts in physical offices, and would require collaboration of other team members, who would be asking why this is higher priority then their current long list of priorities. I would myself struggle to justify it, as the CLI has tens of much higher priority, long standing issues to be resolved...).

Comment (josecoelho) [reply package.json:211]: Can we have another issue to write the missing tests? I might be able to slowly write them up when I get bored :). I believe it's less than ideal to have such a crucial dependency requiring non specified manual tests before an upgrade. It's a big risk to take on if you need to solve a problem that requires and upgrade to the `sdk` quickly.

Comment (pdcastro) [reply package.json:211]: Sure thing! But we cannot merge this PR without at least manually testing the CLI (the non specified manual tests). I was not necessarily expecting you to test the CLI - if you don't, I will. :-)  We need/want to keep the CLI updated with newer SDK versions anyway. Another complication is that openBalena and balenaCloud are not perfectly in sync re API and SDK, and yet we currently have only one CLI that should work for both. I will also ask team members if they are aware of any SDK changes that might behave differently between openBalena and balenaCloud.

Comment (pdcastro) [reply package.json:211]: > _I will also ask team members if they are aware of any SDK changes that might behave differently between openBalena and balenaCloud_

Got an "all clear" from the team re openBalena. It should be enough to test with balenaCloud.

File: lib/errors.ts
Comment (pdcastro) [inline lib/errors.ts:126]: I think this sentence would be better reading something like:
```
Error reading data directory: "${dataDirectory}"
```
... where `dataDirectory` might be obtained with something like:
```
const settings = await import('balena-settings-client');
const dataDirectory = settings.get('dataDirectory') as string;
```
Similar to how it is being done in [events.ts L67](https://github.com/balena-io/balena-cli/blob/v12.27.4/lib/events.ts#L67). Plus some error handling in case obtaining the value of `dataDirectory` fails as well, perhaps for the same reasons as reading the data directory fails (ownership, permissions). In case an error happens attempting to obtain the value of `dataDirectory`, a default value could be used.

I think we should avoid the word "settings" because it is similar to "configuration", and the CLI uses a configuration file which is typically `~/.balenarc.yml` -- see [more details in the FAQ](https://github.com/balena-io/balena-cli/blob/master/TROUBLESHOOTING.md#where-is-the-balena-clis-configuration-file-located). The configuration file and the data directory are different things that may both/either fail independently. The error handling in this PR regards the data directory, not the configuration file.

The location of the data directory is typically `~/.balena` but can be changed using the configuration file (typically `~/.balenarc.yml`) or by using the BALENARC_DATA_DIRECTORY environment variable. More details at:
* https://github.com/balena-io/balena-cli/blob/master/TROUBLESHOOTING.md#how-do-i-make-the-balena-cli-persist-data-in-another-directory
* https://www.balena.io/docs/learn/more/masterclasses/advanced-cli/#13-separate-environment-configurations

By the way, we should test that the `BalenaSettingsPermissionError` is still raised as expected if the data directory is modified using the configuration file or env var.

File: lib/errors.ts
Comment (pdcastro) [inline lib/errors.ts:129]: These 2 lines could be deleted if my suggestion in the other comment is adopted (using `${dataDirectory}` in the first line of the error message). These hardcoded strings for Windows and Linux/macOS could be used as default values for `dataDirectory` if its value cannot be fetched.

File: lib/errors.ts
Comment (pdcastro) [inline lib/errors.ts:132]: _"if the CLI was executed..."_ (or _"if the balena CLI was executed...")_  instead of _"if balena was executed..."_
This is because balena is the name of the company, not the name of the CLI tool. The tool does not yet have its own proper name. It reads "The balena Command Line Interface", which is comparable to saying "The Microsoft Operating System" or "The Microsoft OS" instead of saying "Microsoft Windows".

File: lib/errors.ts
Comment (pdcastro) [inline lib/errors.ts:137]: `${dataDirectory}` as well.

File: package.json
Comment (pdcastro) [reply package.json:211]: I've done some brief manual testing of selected commands that exercise the SDK, using this line in `package.json`:
```
 "balena-sdk": "^15.18.0-josecoelho-update-balena-auth-36282cee7b3f394b690d0d430dc5705c075b67d1",
```
It seem to be all working OK... I was also going through the SDK changelog with a bit more attention, and I was not alarmed by any of the entries. I would be OK with approving this PR, after the proper SDK version is available. I will be chasing the team about the CI failure to generate a new SDK version in its Changelog and npm registry.

**Comment by pdcastro** [comment]
balena SDK v15.18.0 has been released :tada:

File: lib/errors.ts
Comment (josecoelho) [reply lib/errors.ts:126]: I'm going to work on this soon.

**Comment by pdcastro** [comment]
@balena-ci retest

File: lib/errors.ts
Comment (pdcastro) [reply lib/errors.ts:132]: > because balena is [...] not the name of the CLI tool

Well, I can see how one might reasonably disagree, given that the command line reads exactly `balena`. 😁 
Still, at balena (the company) we don't refer to the balena CLI as "balena". 😀 
Btw, on balenaOS devices (host OS), `balena` is a symlink to `balena-engine`, whose name is `balenaEngine`...

**Comment by josecoelho** [comment]
@pdcastro Sorry mate, I must be able to get back to finish it this week. I'm sorry about the delay.

**Comment by pdcastro** [comment]
> @pdcastro Sorry mate, I must be able to get back to finish it this week. I'm sorry about the delay.

That's fine @josecoelho 👍  If you couldn't finish it, I could finish it for you, but I don't want to steal your PR after all the effort you put in! 😁   By the way, the CLI is now using `balena-sdk` v15.20.0 and `balena-errors` v4.7.1, so those changes to `package.json` are no longer needed.

**Comment by josecoelho** [comment]
@pdcastro Requested changes are done. 
I used `CliSettings` form bootstrap instead to avoid dealing with `async` on the error message builder.

I did test it setting `BALENARC_DATA_DIRECTORY` as well. And it works as expected.
I didn't test on windows.


Example of test:

```
$ export BALENARC_DATA_DIRECTORY=~/.balena-other
$ balena login # do login
$ balena apps
# list of apps
$ chmod -R a-r ~/.balena-other
$ balena apps
Error reading data directory: "/home/dev/.balena-other"

This error usually indicates that the user doesn't have permissions over that directory,
which can happen if balena CLI was executed as the root user.

Try resetting the ownership by running: `sudo chown -R $(whoami) /home/dev/.balena-other`

Additional information may be available with the `--debug` flag.

For further help or support, visit:
https://www.balena.io/docs/reference/balena-cli/#support-faq-and-troubleshooting
```

**Comment by pdcastro** [comment]
@josecoelho, looking good! 👍   The build is failing some checks for the shrinkwrap file:
```
** npm-shrinkwrap.json was not deduplicated or not fully committed - FAIL **
** Please run 'npm ci', followed by 'npm dedupe' **
Error: "test-shrinkwrap": Error: /tmp/build/b3c2877b/versioned-source/automation/test-lock-deduplicated.sh failed with exit code 1:
"/tmp/build/b3c2877b/versioned-source/automation/test-lock-deduplicated.sh" [undefined]
```
If you run `npm ci && npm dedupe` as advised by the error message, you still need to run `npm install` as well afterwards:
```
npm ci && npm dedupe && npm install
```
Alternatively, the following also works:
```
git checkout master -- npm-shrinkwrap.json
rm -rf node_modules && npm install
```
But first of all, rebase on the current/latest master branch.

File: lib/errors.ts
Comment (pdcastro) [inline lib/errors.ts:137]: > Example of test:
> [...]
> Additional information may be available with the `--debug` flag.
> For further help or support, visit: [...]

The part above that reads _"Additional information ..."_ indicates that the CLI is treating this error as "unexpected" (as if improperly handled), and reporting it to Sentry.io. Let's avoid this by declaring this error as "expected". I think it is just a matter of adding the following entry to the following array -- find it in this same file:

```
const EXPECTED_ERROR_REGEXES = [
	/^BalenaSettingsPermissionError/,
```

Comment (josecoelho) [reply lib/errors.ts:137]: Sweet. NPM fixed and now it's an expected error.

New Message:
```
Error reading data directory: "/home/dev/.balena"

This error usually indicates that the user doesn't have permissions over that directory,
which can happen if balena CLI was executed as the root user.

Try resetting the ownership by running: `sudo chown -R $(whoami) /home/dev/.balena`
```

**Comment by pdcastro** [comment]
@josecoelho, almost there... 😀 
The build is still failing (same error). Something I told you wasn't right, the following is **not** enough:

```
$ git checkout master -- npm-shrinkwrap.json
$ rm -rf node_modules && npm install
```

I actually had to run the following commands:
```
$ git checkout master -- npm-shrinkwrap.json
$ rm -rf node_modules
$ npm install && npm dedupe && npm install
```

After which, I got the following diff with master:
```
$ git diff master -- npm-shrinkwrap.json

diff --git a/npm-shrinkwrap.json b/npm-shrinkwrap.json
index dc621615..43d5b8c9 100644
--- a/npm-shrinkwrap.json
+++ b/npm-shrinkwrap.json
@@ -2534,18 +2534,6 @@
         "balena-settings-storage": "^7.0.0",
         "jwt-decode": "^2.2.0",
         "tslib": "^2.0.0"
-      },
-      "dependencies": {
-        "balena-settings-storage": {
-          "version": "7.0.0",
-          "resolved": "https://registry.npmjs.org/balena-settings-storage/-/balena-settings-storage-7.0.0.tgz",
-          "integrity": "sha512-gufzVJznyt9e1CvpBuLe2caU5KcEwl1YHCbK5OMz09zXDA2OMAICPXsLlViK+KiuZwZrBx3tyU2FZjAzRZFgwQ==",
-          "requires": {
-            "@types/node": "^10.17.26",
-            "balena-errors": "^4.7.1",
-            "tslib": "^2.0.0"
-          }
-        }
       }
     },
     "balena-config-json": {
@@ -2793,11 +2781,12 @@
       }
     },
     "balena-settings-storage": {
-      "version": "6.0.1",
-      "resolved": "https://registry.npmjs.org/balena-settings-storage/-/balena-settings-storage-6.0.1.tgz",
-      "integrity": "sha512-jdDoKzbJXlF696EZSbwD6lZ1dMe98aUtx7btFE4j4PRCSeh2BWx5P5VLGh9Bk3sH2FUcqYg0iw/wdKvkcv44oA==",
+      "version": "7.0.0",
+      "resolved": "https://registry.npmjs.org/balena-settings-storage/-/balena-settings-storage-7.0.0.tgz",
+      "integrity": "sha512-gufzVJznyt9e1CvpBuLe2caU5KcEwl1YHCbK5OMz09zXDA2OMAICPXsLlViK+KiuZwZrBx3tyU2FZjAzRZFgwQ==",
       "requires": {
         "@types/node": "^10.17.26",
+        "balena-errors": "^4.7.1",
         "tslib": "^2.0.0"
       }
     },
```

Fingers crossed that with the diff above, the build will finally be happy. 🤞

**Comment by josecoelho** [comment]
Done. I've got the same diff as yours. I hope it goes well now. \o/

**Comment by pdcastro** [comment]
> I didn't test on windows.

At all, or just the `BALENARC_DATA_DIRECTORY` variable? I have just tested this branch on Windows, and it did not seem work as expected, although also no worse than the master branch. 😀 

I tested with PowerShell, cmd.exe and MSYS 2. I removed permissions over the `_balena` folder in the home directory, using the Windows Explorer GUI. Results:

On Windows, **before** this PR:
```
PS C:\Users\paulo> dir _balena
dir : Access to the path 'C:\Users\paulo\_balena' is denied.
At line:1 char:1
+ dir _balena
+ ~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (C:\Users\paulo\_balena:String) [Get-ChildItem], UnauthorizedAccessExc
   eption
    + FullyQualifiedErrorId : DirUnauthorizedAccessError,Microsoft.PowerShell.Commands.GetChildItemCommand

PS C:\Users\paulo> (get-command balena).Source
C:\Program Files\balena-cli\bin\balena.cmd

PS C:\Users\paulo> balena version -a
balena-cli version "12.30.2"
Node.js version "10.21.0"

PS C:\Users\paulo> balena apps
Login required: use the “balena login” command to log in.

PS C:\Users\paulo> balena login
 _            _
| |__   __ _ | |  ____  _ __    __ _
| '_ \ / _` || | / __ \| '_ \  / _` |
| |_) | (_) || ||  ___/| | | || (_) |
|_.__/ \__,_||_| \____/|_| |_| \__,_|

Logging in to balena-cloud.com
? How would you like to login? Web authorization (recommended)
Opening web browser for URL:
https://dashboard.balena-cloud.com/login/cli/http%253A%252F%252F127.0.0.1%253A58505%252Fauth
You don't have sufficient privileges to run this operation.
Run a new Command Prompt as administrator and try running this command again.

If this is not the case, and you're trying to burn an SDCard, check that the write lock is not set.

Additional information may be available with the `--debug` flag.

For further help or support, visit:
https://www.balena.io/docs/reference/balena-cli/#support-faq-and-troubleshooting
```

On Windows, **after** this PR:
```
PS C:\Users\paulo> dir _balena
dir : Access to the path 'C:\Users\paulo\_balena' is denied.
At line:1 char:1
+ dir _balena
+ ~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (C:\Users\paulo\_balena:String) [Get-ChildItem], UnauthorizedAccessExc
   eption
    + FullyQualifiedErrorId : DirUnauthorizedAccessError,Microsoft.PowerShell.Commands.GetChildItemCommand

PS C:\Users\paulo> cd .\balena\balena-io\balena-cli\

PS C:\Users\paulo\balena\balena-io\balena-cli> node bin\balena version -a
balena-cli version "12.30.2"
Node.js version "12.20.0"

PS C:\Users\paulo\balena\balena-io\balena-cli> node bin\balena apps
Login required: use the “balena login” command to log in.

PS C:\Users\paulo\balena\balena-io\balena-cli> node bin\balena login
 _            _
| |__   __ _ | |  ____  _ __    __ _
| '_ \ / _` || | / __ \| '_ \  / _` |
| |_) | (_) || ||  ___/| | | || (_) |
|_.__/ \__,_||_| \____/|_| |_| \__,_|

Logging in to balena-cloud.com
? How would you like to login? Web authorization (recommended)
Opening web browser for URL:
https://dashboard.balena-cloud.com/login/cli/http%253A%252F%252F127.0.0.1%253A58487%252Fauth
You don't have sufficient privileges to run this operation.
Run a new Command Prompt as administrator and try running this command again.

If this is not the case, and you're trying to burn an SDCard, check that the write lock is not set.

Additional information may be available with the `--debug` flag.

For further help or support, visit:
https://www.balena.io/docs/reference/balena-cli/#support-faq-and-troubleshooting
```

So it is as if this PR made no difference on Windows...

The difference on macOS:

Before this PR:
```
$ balena version -a
balena-cli version "12.30.2"
Node.js version "12.20.0"

$ balena apps
BalenaSettingsPermissionError: EACCES: permission denied, open '/Users/paulo/.balena/token'

Additional information may be available with the `--debug` flag.

For further help or support, visit:
https://www.balena.io/docs/reference/balena-cli/#support-faq-and-troubleshooting
```

After this PR:
```
$ ./bin/balena apps
Error reading data directory: "/Users/paulo/.balena"

This error usually indicates that the user doesn't have permissions over that directory,
which can happen if balena CLI was executed as the root user.

Try resetting the ownership by running: `sudo chown -R $(whoami) /Users/paulo/.balena`
```

I think it is up to you whether you would like to get the PR merged as it is anyway, or whether you would like to investigate the behaviour on Windows before merging. It is an improvement to merge it as it is anyway!

One final small thing: now that this PR no longer bumps the `balena-sdk` version, and the only user-noticeable effect is a more friendly error message, the `Change-type` footer in the commit could be `patch` instead of `minor`.

**Comment by josecoelho** [comment]
We might need a new issue to handle that on a windows machine. The problem might be on `balena-settings-storage` and the way it captures `EACESS` errors. I'm not sure.

I'm going to update the commit to be `patch`.

**Comment by pdcastro** [comment]
@balena-ci rebase

**Review by pdcastro — APPROVED** [review]

**Comment by pdcastro** [comment]
> balena-ci rebase
> An error occurred while processing the comment. 

Looks like the rebase command of balena-ci still doesn't work with forked repos. @josecoelho, as soon as you rebase the PR on master, tests shall pass again and the PR will get merged. Thank you for you effort and perseverance! 😀

**Comment by josecoelho** [comment]
@pdcastro I'm sorry mate, I forgot about this. I will rebase as soon as I am back home.

**Pull request merged**
